dotnet/src/Connectors/Connectors.AI.OpenAI/ImageGeneration/AzureOpenAIImageGeneration.cs

```diff
@@ -0,0 +1,153 @@
+    /// <param name="maxRetryCount">Maximum number of retries for HTTP requests.</param>
+    /// <param name="defaultRetryAfter">Default retry wait time in seconds for HTTP requests.</param>
+    public AzureOpenAIImageGeneration(string endpoint, string apiKey, HttpClient? httpClient = null, ILogger? logger = null, int maxRetryCount = 5, int defaultRetryAfter = 6) : base(httpClient, logger)
+    {
+        this._endpoint = $"{endpoint}openai/images/generations:submit?api-version=2023-06-01-preview";
+        this._apiKey = apiKey;
+        this._maxRetryCount = maxRetryCount;
+        this._defaultRetryAfter = defaultRetryAfter;
+    }
+
+    /// <inheritdoc/>
+    public async Task<string> GenerateImageAsync(string description, int width, int height, CancellationToken cancellationToken = default)
+    {
+        Verify.NotNull(description);
+        if (width != height || (width != 256 && width != 512 && width != 1024))
+        {
+            throw new ArgumentOutOfRangeException(nameof(width), width, "OpenAI can generate only square images of size 256x256, 512x512, or 1024x1024.");
+        }
+
+        var requestBody = Json.Serialize(new ImageGenerationRequest
+        {
+            Prompt = description,
+            Size = $"{width}x{height}",
+            Count = 1
+        });
+
+        var result = await this.GetImageGenerationAsync(this._endpoint, requestBody, cancellationToken).ConfigureAwait(false);
+        if (result.Result == null)
+        {
+            throw new AIException(AIException.ErrorCodes.InvalidResponseContent, "Response not contains result");
+        }
+
+        return result.Result.Images[0].Url;
+    }
+
+    private async Task<AzureImageGenerationResponse> GetImageGenerationAsync(string url, string? requestBody, CancellationToken cancellationToken = default)
+    {
+        HttpResponseMessage? response = null;
+        var retryCount = 0;
+        try
+        {
+            using (var content = new StringContent(requestBody, Encoding.UTF8, "application/json"))
+            {
+                response = await this.ExecuteRequestAsync(url, HttpMethod.Post, content, cancellationToken).ConfigureAwait(false);
+            }
+
+            var responseJson = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+            var result = this.JsonDeserialize<AzureImageGenerationResponse>(responseJson);
+
+            if (!response.Headers.TryGetValues("operation-location", out var locationValues) || string.IsNullOrWhiteSpace(locationValues.FirstOrDefault()))
+            {
+                throw new AIException(AIException.ErrorCodes.InvalidResponseContent, "Not found operation-location");
+            }
+
+            var operationLocation = locationValues.First();
+            while (true)
+            {
+                if (result.Status.Equals(SucceededStatus, StringComparison.OrdinalIgnoreCase))
+                {
+                    return result;
```

Comment:
There's no need to implement the retry functionality in this class because a default one configured during Kernel building will be provided automatically when the class is registered by the WithAzureOpenAIImageGenerationService method. In case the class is created directly, without using this With* method, it's always possible inject an instance of HttpClient with any retry policy.

Reply:
comments here
https://github.com/microsoft/semantic-kernel/pull/1209#discussion_r1220689587